### DISCUSSION TITLE
MQTT Subscribe strangeness

### DISCUSSION THREAD

**Discussion by msperl** [arendst/Tasmota#18738] [Q&A]
So I have this strange situation:

I do the following subscribe on a selfbuilt tasmota based on devel:

```
Subscribe MSTESTD, tele/nit-d-sr/SENSOR, Time
Subscribe MSTESTS, tele/nit-s-sr-mitte/SENSOR, Time
```

then I have a minimal berry script like this for debug purposes:
```
def myhasp_rule_dump(value, trigger, message)
  import string
  tasmota.log(string.format("XXXXX: %s - %s - %s", value, trigger, message))
end # def myhasp_rule_dump
tasmota.add_rule("Event", myhasp_rule_dump)
```

Strange thing is that for the MSTESTD I get the time of that Tamsmota device printed:
```
14:13:43.980 XXXXX: {'MSTESTD': '2023-05-29T14:13:43'} - Event - {'Event': {'MSTESTD': '2023-05-29T14:13:43'}}
```
(Same if I subscribe to the local topic ("tele/nit-d-test/SENSOR)

But for the MSTESTS I get an actual error in the WebConsole (with WebLog 4):

```
14:12:37.511 EQ3 mqtt: tele/nit-s-sr-mitte/SENSOR:{"Time":"2023-05-29T14:12:36","open":"ON","close":"ON","ENERGY":{"TotalStartTime":"2021-06-30T14:58:57","Total":0.411,"Yesterday":0.000,"Today":0.003,"Period":[0,0],"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":50,"Voltage":235,"Current":[0.000,0.000]},"Shutter1":{"Position":40,"Direction":0,"Target":40,"Tilt":0}}
14:12:37.514 SRC: MQTT
14:12:37.515 CMD: Grp 0, Cmd 'SENSOR', Idx 1, Len 360, Pld -99, Data '{"Time":"2023-05-29T14:12:36","open":"ON","close":"ON","ENERGY":{"TotalStartTime":"2021-06-30T14:58:57","Total":0.411,"Yesterday":0.000,"Today":0.003,"Period":[0,0],"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":50,"Voltage":235,"Current":[0.000,0.000]},"Shutter1":{"Position":40,"Direction":0,"Target":40,"Tilt":0}}'
14:12:37.521 MQT: stat/nit-d-test/RESULT = {"Command":"Error"}
```

All 3 devices have different versions:
* nit-d-test: Device subscribing: 12.5.0.3 (dev build  - custom tasmota32 for lanbon-l8 display - needed because of PSRAM not detected with default builds)
* nit-d-sr: 12.5.0 (release build - custom tasmota32 for lanbon-l8 display)
* nit-s-sr-mitte: 12.4.0 (release build - standard tasmota32 for a shutter config on a Sonoff Dual R3)

Any ideas why these - in principle - identical subscribes behaves differently?

**Comment by sfromis** [comment]
Better just not use the old `Subscribe` command and `Event` for Berry, but instead `mqtt.subscribe`

Processing MQTT without flagging for complete tends to lead to it to be passed to next step in the flow, meaning as a command.

**Reply by msperl** [reply]
Well, it is for haspmota using val_rule - and so `tasmota.cmd("subscribe ...")` is quicker/shorter (and it should work...)

Not sure how I would implement the same thing using berry mqtt.subscribe instead with the same minimum amount of coding...

**Comment by msperl** [comment]
It is definitely strange!

I have upgraded the nit-s-sr-mitte node to also 12.5.0.4 devel and it gives the same error when subscribing to its own topic, but it does not give a hickup when subscribing to nit-d-sr or nit-d-test:

```
16:07:54.171 CMD: Subscribe MSTESTS, tele/nit-s-sr-mitte/SENSOR, Time
16:07:54.192 MQT: stat/nit-s-sr-mitte/RESULT = {"Subscribe":"MSTESTS,tele/nit-s-sr-mitte/SENSOR/#,Time"}
16:08:09.548 MQT: tele/nit-s-sr-mitte/SENSOR = {"Time":"2023-05-29T16:08:09","open":"ON","close":"ON","ENERGY":{"TotalStartTime":"1970-01-01T00:00:00","Total":0.411,"Yesterday":0.000,"Today":0.003,"Period":[0,0],"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":50,"Voltage":236,"Current":[0.000,0.000]},"Shutter1":{"Position":40,"Direction":0,"Target":40,"Tilt":0}}
16:08:09.585 MQT: stat/nit-s-sr-mitte/RESULT = {"Command":"Error"}
16:08:27.115 CMD: Subscribe MSTESTS, tele/nit-d-sr/SENSOR, Time
16:08:27.136 MQT: stat/nit-s-sr-mitte/RESULT = {"Subscribe":"MSTESTS,tele/nit-d-sr/SENSOR/#,Time"}
```
(without debug logging enabled)

**Reply by msperl** [reply]
Is maybe the mqtt json too big? (just a thought because I just found out that OTAURL is also length limited)

**Reply by msperl** [reply]
OK, so If i just publish something like this to that topic:
```
{ "Time": "2023-05-29T16:19:09" }
```
Then it works!

So is there some sort of limit somewhere in the mqtt json parser?

For comparison the json size:
* Working (displays): 193 chars
* not Working: 360 chars

I believe tasmota should be able to consume what it produces...

Is there a define to set that buffer size when self-compiling tasmota?

**Reply by sfromis** [reply]
Did you try the better way I suggested? Illustrative example:

```
def myhasp_rule_dump(topic, idx, payloadstr, databytes)
  import json
  var payloadmap = json.load(payloadstr)
  print(topic, payloadmap.find('Time', nil), payloadmap)
  return true  # this is to indicate that the payload is not to be processed further
end # def myhasp_rule_dump
import mqtt
mqtt.subscribe('+/tele/SENSOR', myhasp_rule_dump)mqtt.subscribe('tele/nit-d-sr/SENSOR', myhasp_rule_dump)
mqtt.subscribe('tele/nit-s-sr-mitte/SENSOR', myhasp_rule_dump)
```

**Reply by arendst** [reply]
> Is there a define to set that buffer size when self-compiling tasmota?

``#define RULE_MAX_MQTT_EVENTSZ   256``

There is a reason why it is so small. Happy debugging when enlarging it.

**Reply by msperl** [reply]
No, that I have not tried.

But would that work with a pages.jsonl file along those lines?
```
{ "page": 255, "id": 5,  "obj": "label", "x": 130,  "y": 96, "w": 100, "h": 32, "radius": 0, "border_side": 0, "text_font": "montserrat-28", "text_color":"#FF0000", "text": "", "text_rule":"Event#MSTESTD", "text_rule_format":"%s"}
```

So I am not sure how that mqtt.subscribe would integrate with haspmota!

The berry thing is mostly my debug option to see what events are produced (which can be handy...) and why it does not show it in the haspmota UI!

As said: I guess/fear that there is a limit to the mqtt parser/buffer that might be set to 256 bytes...

**Reply by msperl** [reply]
OK, but then there should be a more direct way for haspmota to render MQTT data that does not rely rules.

At the very least: maybe an error message that says: `MQTT size is limited to 256 bytes - can not process ...` - and not that  error message `{"Command":"Error"}` that does not really give an indication where things go wrong...

**Reply by msperl** [reply]
When the haspmota "meta" field finally lands in a stable release, then I might be tempted to implement a module that does some discovery and handle direct mqtt subscription and rendering of those mqtt values...

Something like this:
```{ ... "meta": { "mqtt_value": { "topic": "...", "path": "...", "compute": "...", "format": "..." } } ... }```

could then probably get implemented fairly easily with simple autodiscovery...

(Trickiest portion is possibly the "path" navigation in the parsed json)

**Reply by sfromis** [reply]
HASPmota rendering of MQTT data is already "more direct", by using Berry MQTT handling instead of rules. On ESP32, better just forget about  the old `Subscribe` command, the more as this is not at all relevant to HASPmota. If you want debugging, using "the Berry way" is representative, which the old non-Berry subscribe is not. You are only adding confusion by mixing the two ways.

While the word "rules" is also used with Berry, this is unrelated to classic Tasmota non-Berry rules, for which Subscribe/Event is the relevant strategy; with Berry the relevant strategy is using `mqtt.subscribe`.

**Reply by arendst** [reply]
> At the very least: maybe an error message that says: MQTT size is limited to 256 bytes - can not process ... - and not that error message {"Command":"Error"} that does not really give an indication where things go wrong...

If we had an abundance of code space we could tell users the world when things go wrong. We don't hence this is what you get.

**Reply by sfromis** [reply]
In this case, just follow my recommendation. Berry `mqtt.subscribe` can handle long payloads without that type of issues, and is already the way used internally on HASPmota.

**Reply by msperl** [reply]
> HASPmota rendering of MQTT data is already "more direct", by using Berry MQTT handling instead of rules. On ESP32, better just forget about the old `Subscribe` command, the more as this is not at all relevant to HASPmota. If you want debugging, using "the Berry way" is representative, which the old non-Berry subscribe is not. You are only adding confusion by mixing the two ways.
> 
> While the word "rules" is also used with Berry, this is unrelated to classic Tasmota non-Berry rules, for which Subscribe/Event is the relevant strategy; with Berry the relevant strategy is using `mqtt.subscribe`.

Maybe I am just missing something, but all I find are the examples in the documentation for haspmota (https://tasmota.github.io/docs/HASPmota) that make use of rules - most specifically an example with “esp32#Temperature”, which does not exist any longer…

so maybe just some hint on how to do it with mqtt directly would be helpful.

as said: my approach would use the meta field and auto Discovery of objects that need it.
(But that would only happen after my vacation)

**Reply by sfromis** [reply]
No, none of the HASPmota examples use rules *in the meaning of "classic" Tasmota rules*. They are *all* based on Berry (which confusingly also use the term rules for a similar type of triggers). If/when you use `ESP32#Temperature` (this bad number can be enabled if you want, SetOption146), this is not classic Tasmota rules, but done in Berry. Thus, you already have the HASPmota examples of using Berry directly, as this is the *only* way to do it from HASPmota.

Hence, I suggest that you do the debugging with the Berry way too, and simply forget about using the old strategy of the Subscribe command + Event, as this is just not representative of that you can do with HASPmota/Berry.

Beware that for debugging, the `return true` from my example should probably not be used, as this may stop HASPmota from getting the same data.

However, to avoid confusion, let me add that if using `mqtt.subscribe` to get data from another device than where it is running, this is not representative of what you can do from HASPmota. The most direct equivalent of HASPmota is using `tasmota.add_rule` to get at the data, without involving MQTT at all. I only pointed at using `mqtt.subscribe` because this is the better way of doing it, instead of using the Subscribe command + Event.

**Reply by msperl** [reply]
Yes, that is probably true, but it would be easier, if the haspmota documentation would show how to present a mqtt value.

The way you write it it seems as if all that is needed is a mqtt.subscribe(“my/topic”) and some way to reference that topic in  the jsonl file (value_rule maybe?), but that is where my disconnect starts and where the documentation does not help solve it!

My best guess now would be to put that topic name subscribed as value_rule as well.

But that is now inferred from this discussion here and not taken from the documentation, so others will also fall into this trap!

**Reply by sfromis** [reply]
As HASPmota has no way of subscribing to MQTT topics for for picking values, you cannot make such references in pages.jsonl. Such needs will have to be solved by other means, like creating Berry code to subscribe to MQTT, and push the values to HASPmota.

**Comment by msperl** [comment]
Well, that is why the simple subscribe is so much easier - as long as it works ;)

and as I have said: I guess I will need to try to use the meta field to reduce the coding to a minimum. If I come wirh something easily reusable, I shall share it…

The basic proposal above would be acceptable?
